etcd-v2/docker-compose.yml

```diff
@@ -0,0 +1,98 @@
+          - ./data1/:/var/lib/etcd/
+          - ./certs/:/etc/ssl/certs/
+        networks:
+          etcd_dev:
+            ipv4_address: 172.30.66.10
+        environment:
+            ETCD_NAME: etcd-1
+            ETCDCTL_API: 3
+            ETCD_DEBUG: 1
+            ETCD_DATA_DIR: /var/lib/etcd
+            ETCD_INITIAL_ADVERTISE_PEER_URLS: http://etcd-1:2380
+            ETCD_INITIAL_CLUSTER: etcd-1=http://etcd-1:2380,etcd-2=http://etcd-2:2380,etcd-3=http://etcd-3:2380
+            ETCD_INITIAL_CLUSTER_STATE: new
+            ETCD_INITIAL_CLUSTER_TOKEN: etcd-ftw
+            ETCD_LISTEN_CLIENT_URLS: http://0.0.0.0:2379
+            ETCD_LISTEN_PEER_URLS: http://0.0.0.0:2380
+            ETCD_ADVERTISE_CLIENT_URLS: http://etcd-1:2379,http://127.0.0.1:2379
+    etcd-2:
+        image: gcr.io/etcd-development/etcd:v3.3.12
+        restart: always
+        ports:
+            - 2379
+            - 2380
+        volumes:
+          - ./certs/:/etc/ssl/certs/
+        networks:
+          etcd_dev:
+            ipv4_address: 172.30.66.11
+        environment:
+            ETCD_NAME: etcd-2
+            ETCDCTL_API: 3
+            ETCD_DEBUG: 1
+            ETCD_DATA_DIR: /var/lib/etcd
+            ETCD_INITIAL_ADVERTISE_PEER_URLS: http://etcd-2:2380
+            ETCD_INITIAL_CLUSTER: etcd-1=http://etcd-1:2380,etcd-2=http://etcd-2:2380,etcd-3=http://etcd-3:2380
+            ETCD_INITIAL_CLUSTER_STATE: new
+            ETCD_INITIAL_CLUSTER_TOKEN: etcd-ftw
+            ETCD_LISTEN_CLIENT_URLS: http://0.0.0.0:2379
+            ETCD_LISTEN_PEER_URLS: http://0.0.0.0:2380
+            ETCD_ADVERTISE_CLIENT_URLS: http://etcd-2:2379,http://127.0.0.1:2379
+    etcd-3:
+        image: gcr.io/etcd-development/etcd:v3.3.12
+        restart: always
+        ports:
+            - 2379
+            - 2380
+        volumes:
+          - ./certs/:/etc/ssl/certs/
+        networks:
+          etcd_dev:
+            ipv4_address: 172.30.66.12
+        environment:
+            ETCD_NAME: etcd-3
+            ETCDCTL_API: 3
+            ETCD_DEBUG: 1
+            ETCD_DATA_DIR: /var/lib/etcd
+            ETCD_INITIAL_ADVERTISE_PEER_URLS: http://etcd-3:2380
+            ETCD_INITIAL_CLUSTER: etcd-1=http://etcd-1:2380,etcd-2=http://etcd-2:2380,etcd-3=http://etcd-3:2380
+            ETCD_INITIAL_CLUSTER_STATE: existing
+            ETCD_INITIAL_CLUSTER_TOKEN: etcd-ftw
+            ETCD_LISTEN_CLIENT_URLS: http://0.0.0.0:2379
+            ETCD_LISTEN_PEER_URLS: http://0.0.0.0:2380
+            ETCD_ADVERTISE_CLIENT_URLS: http://etcd-3:2379,http://127.0.0.1:2379
+
+    etcd-client:
+        image: gcr.io/etcd-development/etcd:v3.3.12
+        restart: always
+        volumes:
+          - ./certs/:/etc/ssh/certs/
+        networks:
+          etcd_dev:
+            ipv4_address: 172.30.66.13
+        environment:
+          ETCDCTL_API: 3
+          ETCDCTL_ENDPOINTS: http://etcd-1:2379,http://etcd-2:2379,http://etcd-3:2379
+
```

Comment:
nice can you tell me how DNS works here? Meaning how does `etcd-1`resolve?

/cc @retroflexer

Reply:
Yeah sure. Docker creates a new local cluster names etcd-dev. You can actually do a `docker inspect network etcd-dev` and it should give the runtime config for this network. 

This puts all the 4 containers in the same bridge. The DNS gets resolved via hostname. Quoting docker-compose documentation `Each container can now lookup the hostname web or db and get back the appropriate container’s IP address.` This is achieved by docker   configuring 3 files inside the container describe [here](https://docs.docker.com/v17.09/engine/userguide/networking/default_network/configure-dns/)